<?xml version="1.0"?>

<launch>
  <node name="ydlidar_node"  pkg="ydlidar"  type="ydlidar_node" output="screen">
<!-- JJ - test ydlidar direct connect to serial port -->
<!--    <param name="port"         type="string" value="/dev/ttyUSB0"/> -->
    <param name="port"         type="string" value="/dev/ttyS0"/>
<!-- JJ - deviate from document recommended value, as lidar appears to report faster connection speed 
    <param name="baudrate"     type="int"    value="153600"/>
    <param name="baudrate"     type="int"    value="115200"/>
    <param name="baudrate"     type="int"    value="230400"/> -->
    <param name="baudrate"     type="int"    value="230400"/>
<!-- JJ - changed to match what we've got in the urdf (base_laser)!
    <param name="frame_id"     type="string" value="laser_frame"/> -->
    <param name="frame_id"     type="string" value="base_laser"/>    
    <param name="angle_fixed"  type="bool"   value="true"/>
    <param name="low_exposure"  type="bool"   value="false"/>
    <param name="heartbeat"    type="bool"   value="false"/>
    <param name="resolution_fixed"    type="bool"   value="true"/>
    <param name="angle_min"    type="double" value="-180" />
    <param name="angle_max"    type="double" value="180" />
    <param name="range_min"    type="double" value="0.08" />
    <param name="range_max"    type="double" value="16.0" />
    <param name="ignore_array" type="string" value="" />
    <param name="samp_rate"    type="int"    value="9"/>
    <param name="frequency"    type="double" value="7"/>
  </node>

  <arg name="tf_map_scanmatch_transform_frame_name" default="scanmatcher_frame"/>
  <arg name="base_frame" default="base_footprint"/>

<!-- JJ
  <arg name="odom_frame" default="nav"/> -->
  <arg name="odom_frame" default="base_frame"/>

  <arg name="pub_map_odom_transform" default="true"/>
  <arg name="scan_subscriber_queue_size" default="5"/>
  <arg name="scan_topic" default="scan"/>
  <arg name="map_size" default="2048"/>
  
  <node pkg="hector_mapping" type="hector_mapping" name="hector_mapping" output="screen">
    
    <!-- Frame names -->
    <param name="map_frame" value="map" />
    <param name="base_frame" value="$(arg base_frame)" />
    <param name="odom_frame" value="$(arg odom_frame)" />
    
    <!-- Tf use -->
    <param name="use_tf_scan_transformation" value="true"/>
    <param name="use_tf_pose_start_estimate" value="false"/>
    <param name="pub_map_odom_transform" value="$(arg pub_map_odom_transform)"/>
    
    <!-- Map size / start point -->
    <param name="map_resolution" value="0.050"/>
    <param name="map_size" value="$(arg map_size)"/>
    <param name="map_start_x" value="0.5"/>
    <param name="map_start_y" value="0.5" />
    <param name="map_multi_res_levels" value="2" />
    
    <!-- Map update parameters -->
    <param name="update_factor_free" value="0.4"/>
    <param name="update_factor_occupied" value="0.9" />    
    <param name="map_update_distance_thresh" value="0.4"/>
    <param name="map_update_angle_thresh" value="0.06" />
    <param name="laser_z_min_value" value = "-1.0" />
    <param name="laser_z_max_value" value = "1.0" />
    
    <!-- Advertising config --> 
    <param name="advertise_map_service" value="true"/>
    
    <param name="scan_subscriber_queue_size" value="$(arg scan_subscriber_queue_size)"/>
    <param name="scan_topic" value="$(arg scan_topic)"/>
    
    <!-- Debug parameters -->
    <!--
      <param name="output_timing" value="false"/>
      <param name="pub_drawings" value="true"/>
      <param name="pub_debug_output" value="true"/>
    -->
    <param name="tf_map_scanmatch_transform_frame_name" value="$(arg tf_map_scanmatch_transform_frame_name)" />
  </node>
 
<!-- JJ - added -->
  <node pkg="tf" type="static_transform_publisher" name="base_frame_broadcaster" args="0.0 0.0 0.0 0.0 0.0 0.0 base_frame base_footprint 100" />
  <node pkg="tf" type="static_transform_publisher" name="base_laser_broadcaster" args="0.0 0.0 0.0 0.0 0.0 0.0 base_footprint laser_frame 100" />


  <node pkg="mini_turty3" type="mini_turty3" name="mini_turty3">
    <param name="enable_odom_transform" value="0"/>
  </node>

<!-- JJ - added robot model -->
  <arg name="model" />
  <arg name="gui" default="False" />
<!--  <param name="robot_description" textfile="$(arg model)" />-->
  <param name="robot_description" textfile="$(find mini_turty3)/launch/mini_turty.urdf" />
  <param name="use_gui" value="$(arg gui)"/>
  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" ></node>
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" ></node>
   
  <!--<node pkg="tf" type="static_transform_publisher" name="map_nav_broadcaster" args="0 0 0 0 0 0 map nav 100"/>-->
</launch>
  
  
